File: chapitre5/exercice3.py
import threading
import random
import time

class TypeTableau:
    def __init__(self):
        self.n = 0
        self.t = []
        self.entier = 0

result = None

def generation(tab):
    print("\n Génération du tableau...")
    tab.t = [random.randint(0, 99) for _ in range(tab.n)]
    time.sleep(2)
    print("[Thread 1] Génération terminée.")

def t_recherche(tab):
    global result
    trouve = 0
    for i in range(tab.n):
        if tab.t[i] == tab.entier:
            trouve = 1
            break
    result = trouve

def main():
    global result
    data = TypeTableau()

    nb_elements = int(input("Combien d'éléments pour le tableau ? "))
    data.n = nb_elements

    t1 = threading.Thread(target=generation, args=(data,))
    t1.start()
    t1.join()

    data.entier = int(input("Pendant ce temps, entrez la valeur x à rechercher : "))

    recherche = threading.Thread(target=t_recherche, args=(data,))
    recherche.start()
    recherche.join()

    if result == 1:
        print(f"L'entier {data.entier} a été trouvé dans le tableau.")
    else:
        print(f"L'entier {data.entier} n'est pas dans le tableau.")

if __name__ == "__main__":
    main()

File: chapitre5/exercice3.c
#include <pthread.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <unistd.h> 

typedef struct
{
    int n;
    int *t;
    int entier; 
} TypeTableau;

void *generation(void *arg)
{
    TypeTableau *tab = (TypeTableau *)arg;
    printf("\n Génération du tableau...\n");
    for(int i = 0; i < tab->n; i++)
    {
        tab->t[i] = rand() % 100;
    }
    sleep(2); 
    printf("[Thread 1] Génération terminée.\n");
    pthread_exit(NULL);
}

void *t_recherche(void *arg)
{
    TypeTableau *tab = (TypeTableau *)arg;
    long trouve = 0; 
    
    for(int i = 0; i < tab->n; i++)
    { 
        if(tab->t[i] == tab->entier)
        {
            trouve = 1;
            break;
        }
    }
    pthread_exit((void*)trouve);
}

int main()
{
    pthread_t t1, recherche;
    TypeTableau data;
    int nb_elements;
    void* status; 

    srand(time(NULL));

    printf("Combien d'éléments pour le tableau ? ");
    scanf("%d", &nb_elements);
    
    data.n = nb_elements;
    data.t = malloc(nb_elements * sizeof(int));

    pthread_create(&t1, NULL, generation, (void*)&data); 
    pthread_join(t1, NULL);

    printf("Pendant ce temps, entrez la valeur x à rechercher : ");
    scanf("%d", &data.entier);

    pthread_create(&recherche, NULL, t_recherche, (void*)&data);
    
    pthread_join(recherche, &status);
    int result = (int)(long)status; 

    if(result == 1)
    {
        printf("L'entier %d a été trouvé dans le tableau.\n", data.entier);
    }
    else
    {
        printf("L'entier %d n'est pas dans le tableau.\n", data.entier);
    }

    free(data.t);
    return 0;
}